# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: accessible/xul/XULComboboxAccessible.cpp
# Commit: 1fbca7956efd
# Full Hash: 1fbca7956efd1b73ae3cc50f21ac2780a9e2040c
# Author: Morgan Rae Reschenberg <mreschenberg@mozilla.com>
# Date: 2025-04-10 09:29:04
# Regressor Bug: 1549202
# File Overlap Count: 1
# Description:
#   Bug 1549202: Ensure ContentSelectDropdown local accessibles return the focused <select> remote accessible as their parent r=Jamie
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D243029
# ==============================================================================

diff -r 0b9e529a5541 -r 1fbca7956efd accessible/xul/XULComboboxAccessible.cpp
--- a/accessible/xul/XULComboboxAccessible.cpp	Wed Apr 09 23:51:01 2025 +0000
+++ b/accessible/xul/XULComboboxAccessible.cpp	Wed Apr 09 23:51:27 2025 +0000
@@ -9,6 +9,9 @@
 #include "nsAccessibilityService.h"
 #include "DocAccessible.h"
 #include "nsCoreUtils.h"
+#include "nsFocusManager.h"
+
+#include "mozilla/a11y/DocAccessibleParent.h"
 #include "mozilla/a11y/Role.h"
 #include "States.h"
 
@@ -140,3 +143,46 @@
 
   return false;
 }
+
+////////////////////////////////////////////////////////////////////////////////
+// XULContentSelectDropdownAccessible
+////////////////////////////////////////////////////////////////////////////////
+
+Accessible* XULContentSelectDropdownAccessible::Parent() const {
+  // We render the expanded dropdown for <select>s in the parent process
+  // as a child of the application accessible. This confuses some
+  // ATs which expect the select to _always_ parent the dropdown (in
+  // both expanded and collapsed states).
+  // To rectify this, we spoof the <select> as the parent of the
+  // expanded dropdown here. Note that we do not spoof the child relationship.
+
+  // First, try to find the select that spawned this dropdown.
+  // The select that was activated does not get states::EXPANDED, but
+  // it should still have focus.
+  Accessible* focusedAcc = nullptr;
+  if (auto* focusedNode = FocusMgr()->FocusedDOMNode()) {
+    // If we get a node here, we're in a non-remote browser.
+    DocAccessible* doc =
+        GetAccService()->GetDocAccessible(focusedNode->OwnerDoc());
+    focusedAcc = doc->GetAccessible(focusedNode);
+  } else {
+    nsFocusManager* focusManagerDOM = nsFocusManager::GetFocusManager();
+    dom::BrowsingContext* focusedContext =
+        focusManagerDOM->GetFocusedBrowsingContextInChrome();
+
+    DocAccessibleParent* focusedDoc =
+        DocAccessibleParent::GetFrom(focusedContext);
+    MOZ_ASSERT(focusedDoc && focusedDoc->IsDoc(), "No focused document found");
+    focusedAcc = focusedDoc->AsDoc()->GetFocusedAcc();
+  }
+
+  if (!NS_WARN_IF(focusedAcc && focusedAcc->IsHTMLCombobox())) {
+    // We can sometimes get a document here if the select that
+    // this dropdown should anchor to loses focus. This can happen when
+    // calling AXPressed on macOS. Call into the regular parent
+    // function instead.
+    return LocalParent();
+  }
+
+  return focusedAcc;
+}